api/buckets/service.go

```diff
@@ -616,36 +739,74 @@ func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (*pb.SetP
 	}
 	dbToken, _ := thread.TokenFromContext(ctx)
 
-	buck := &tdb.Bucket{}
-	err := s.Buckets.Get(ctx, dbID, req.Key, buck, tdb.WithToken(dbToken))
+	destPath := cleanPath(req.Path)
+	bootCid, err := cid.Decode(req.Cid)
 	if err != nil {
+		return nil, fmt.Errorf("invalid remote cid: %s", err)
+	}
+
+	buck := &tdb.Bucket{}
+	if err = s.Buckets.GetSafe(ctx, dbID, req.Key, buck, tdb.WithToken(dbToken)); err != nil {
 		return nil, fmt.Errorf("get bucket: %s", err)
 	}
-	buckPath := path.New(buck.Path)
 
-	remoteCid, err := cid.Decode(req.Cid)
+	buck.UpdatedAt = time.Now().UnixNano()
+	buck.SetMetadataAtPath(destPath, tdb.Metadata{
+		UpdatedAt: buck.UpdatedAt,
+	})
+	buck.UnsetMetadataWithPrefix(destPath + "/")
+
+	txn, err := s.Buckets.WriteTxn(ctx, dbID, tdb.WithToken(dbToken))
 	if err != nil {
-		return nil, fmt.Errorf("invalid remote cid: %s", err)
+		return nil, err
+	}
+	if err = txn.Verify(ctx, buck); err != nil {
+		return nil, err
 	}
-	remotePath := path.IpfsPath(remoteCid)
 
-	encKey := buck.GetEncKey()
-	var dirpth path.Resolved
-	if req.Path == "" {
-		sn, err := makeSeed(encKey)
+	var linkKey, fileKey []byte
+	if buck.IsPrivate() {
+		linkKey = buck.GetLinkEncryptionKey()
+		fileKey, err = buck.GetFileEncryptionKeyForPath(destPath)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	buckPath := path.New(buck.Path)
+	dirPath, err := s.setPathFromExistingCid(ctx, buck, buckPath, destPath, bootCid, linkKey, fileKey)
+	if err != nil {
+		return nil, err
+	}
+	buck.Path = dirPath.String()
+	if err = txn.Save(ctx, buck); err != nil {
+		return nil, err
+	}
+	if err = txn.End(); err != nil {
```

Comment:
Here's the final commit.

Reply:
Oh yeah, good catch. The error from ending the transaction is important... I went with this style,
```
txn, err := s.Buckets.WriteTxn(ctx, dbID, tdb.WithToken(dbToken))
if err != nil {
	return nil, err
}
defer func() {
	if e := txn.End(); err == nil {
		err = e
	}
}()
```
Where `err` is named param in the surrounding function.

Reply:
But what indicates that End will commit or rollback?
On an early return or in a success case both will call `End` without any extra thing.

Reply:
```
	if e := txn.End(); err == nil {
		err = e
	}
```
Is that `== nil` intended, or `!= nil`?

Reply:
`End` will commit if there are things to commit, which in this case only happens if `Save` below succeeds. 

The `err` check is saying, if the error isn't already set (from outside the defer), return the one that happened here.

Reply:
Ah, was `err == nil` not `e == nil`. :)

Regarding `End`: but what happens if you have two `Save(...)`, first succeeds, and then there's an early return before the second `Save(..)` call. You don't want the first `Save(...)` to be committed right?

Reply:
Ah yeah, that would be bad. Just to be clear about the correct usage, I could add a manual `Discard` to the error catch when using save.

Reply:
Cool, yeah. So the client should call `Discard` in any kind of early return while the transaction is open.
Might worth adding to the "WriteTxn(..)" comments 👍🏼

Reply:
Hrm, looks like I need to add `Discard` to the DB API.